### PR TITLE
TV divergence Rho filtering

### PR DESCRIPTION
**Pull request by mnoukhov**
Mask gradients according to TVPO (my name for the LLM-modified version of VACO https://arxiv.org/abs/2603.01365)

Still maintain `rho` as the multiplier but change how we calculate which tokens to filter

Mask a token it exceeds the sequence-level TV divergence `D_tv = |rho - 1|` exceeds the set bounds **and** that token will purportedly increase the TV divergence if a gradient is applied `advantage * sign(logprob - vllm_logprob) > 0`

Algo from the paper:

image

### CODE REVIEW

File: open_instruct/grpo_utils.py [hunk 1]
```diff
@@ -120,6 +120,11 @@ class GRPOExperimentConfig(
     compute the mean log-ratio (1/|o_i|) Σ_t log(π_old / π_θ) per response sequence,
     exponentiate to get a per-sequence ρ, and broadcast the keep/drop decision to every
     token in that sequence. If False (default), the mask is applied per token."""
+    rho_mask_tv_divergence: bool = False
+    """If True, applies the TV divergence masking from VACO (https://arxiv.org/abs/2603.01365)
+    maintains same rho for correction but masks using bounds and sequence-level TV divergence (1 - abs(ρ))
+    won't mask tokens that purport to decrease TV divergence: advantage * ρ <= 0
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/grpo_utils.py]: fixed

File: open_instruct/grpo_utils.py [hunk 2]
```diff
@@ -401,9 +410,21 @@ def compute_rho_correction(
         return RhoCorrection(weights=torch.ones_like(rho), metrics={}, histogram_metrics=rho_hist)
 
     rho_effective = _rho_sequence_level(logprob_diff, response_mask) if config.rho_mask_sequence_level else rho
-    dropped_low, dropped_high = _rho_drop_masks(
-        rho_effective, response_mask, config.rho_mask_lower_bound, config.rho_mask_upper_bound
-    )
+
+    if config.rho_mask_tv_divergence:
+        tv_divergence = torch.abs(rho - 1.0)
+        tv_sequence_level = _rho_sequence_level(tv_divergence, response_mask)
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/grpo_utils.py]: suggestion is wrong but the issue of rho_sequence_level is real, fixed

File: open_instruct/grpo_utils.py [hunk 3]
```diff
@@ -126,6 +126,11 @@ class GRPOExperimentConfig(
     compute the mean log-ratio (1/|o_i|) Σ_t log(π_old / π_θ) per response sequence,
     exponentiate to get a per-sequence ρ, and broadcast the keep/drop decision to every
     token in that sequence. If False (default), the mask is applied per token."""
+    rho_mask_tv_divergence: bool = False
+    """If True, applies the TV divergence masking from VACO (https://arxiv.org/abs/2603.01365)
+    maintains same rho for correction but masks using bounds and sequence-level TV divergence abs(p_seq - 1)
+    won't mask tokens that purport to decrease TV divergence: advantage * ρ <= 0
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/grpo_utils.py]: fixed docstring

**Review by finbarrtimbers — APPROVED** [review]
Thanks for your patience!
